### PR TITLE
feat: use storage mounts to keep config and db files

### PR DESCRIPTION
**Pull request by kayra1**
# Description
This PR introduces 3 things:

* Moving to a central configure function
* Using storage mounts for configuration files and the database.
* Generating a self signed certificate automatically on charm deployment.

I combined these 3 changes because they were heavily related to each other and were required to have a successfully running charm.

There are several TODO comments in the code where logic will need to be written. These are left explicitly for different PR's for 2 reasons:
* This PR is a nice chunk of code relevant to starting GoCert properly.
* They depend on logic for creating the first user which is out of scope of this PR.

## Parametrized Unit Testing
I tried to take a different approach to the tests we commonly write for charms by heavily utilizing pytest's parametrization feature. The format I decided on was to have a test for each public handler, and parametrize all of the charm state independently. 

The parameters currently introduced are only Storage, Container, Network and GoCert. 
The first value in a parameter is either:
* A mock to represent the resource that is being parametrized
* A scenario object like `Storage` that will be automatically mocked when passed to the state`.
This value is what is passed into the state before the event is run

The second value in a parameter represents the status of the object. This status could be anything, but in this case it's either a boolean to indicate readyness or a literal string representing the status of the mock.
This value is used to determine the expected output.

### Why?
Charms have exponentially increasing input states, but the expected transition from one state to the other is usually limited in number. In this case at this point, 36 different possible input states can only lead to 1 of 7 output states of the charm. A significant part of charm development is handling edge cases in the input state to make sure that all possible input states lead to the same expected output state, which this format of writing tests categorically enhance. 

Parametrizing all possible states of the different state objects and having pytest automatically attempt all of them simultaneously helps catch these edge cases while reducing the amount of repetitive tests that users have to write. Pytest may even create tests that users haven't thought of and catch exceptions that are very rare, because multiple parametrize decorators create an exhaustive matrix of all state objects.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -5,80 +5,160 @@
...
-    def _install(self, event: ops.PebbleReadyEvent):
+    def configure(self, event: ops.EventBase):
+        """Handle configuration events."""
+        if not self.unit.is_leader():
+            return
+        if not self.container.can_connect():
+            return
+        self._configure_gocert_config_file()
+        self._configure_access_certificates()
+        self._configure_certificate_relations()
```
Comment (gruyaume) [inline src/charm.py:81]: Pretty clean!

File: src/charm.py [hunk 2]
```diff
@@ -95,6 +175,117 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+
+    @property
+    def _certificate_requests_table(self) -> List[dict[str, str]] | None:
+        # TODO: Error checking
+        r = requests.get(
+            url=f"https://{self._application_bind_address}:{self.port}/api/v1/certificate_requests",
+            verify=f"{CHARM_PATH}/{CONFIG_MOUNT}/0/ca.pem",
+        )
+        gocert_csrs = r.json()
+        return gocert_csrs
```
Comment (gruyaume) [inline src/charm.py:197]: I'd encourage to include the error checking in the same PR

File: src/charm.py [hunk 3]
```diff
@@ -95,6 +175,117 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+        return True
+
+    def _gocert_initialized(self) -> bool:
+        """Return if gocert is initialized."""
+        try:
+            req = requests.get(
+                f"https://{self._application_bind_address}:{self.port}/status",
+                verify=f"{CHARM_PATH}/{CONFIG_MOUNT}/0/ca.pem",
+            )
+        except (requests.RequestException, OSError):
```
Comment (gruyaume) [inline src/charm.py:224]: Why OSError?

File: src/charm.py [hunk 4]
```diff
@@ -5,80 +5,160 @@
...
-    def _on_collect_status(self, event: ops.CollectStatusEvent):
-        event.add_status(ops.ActiveStatus())
+    def _configure_access_certificates(self):
+        """Update the config files for gocert and replan if required."""
+        logger.info("[GoCert] Configuring certificates.")
+        certificates_changed = False
+        if not self._self_signed_certificates_generated():
+            certificates_changed = True
+            self._generate_self_signed_certificates()
+        logger.info("[GoCert] Certificates configured.")
```
Comment (gruyaume) [inline src/charm.py:148]: Why the [GoCert] prefix everywhere? Juju will tell us the app/unit that triggered the log

File: src/charm.py [hunk 5]
```diff
@@ -5,80 +5,160 @@
...
+            return
+        if not self._gocert_available():
+            event.add_status(ops.WaitingStatus("GoCert server not yet available"))
+            return
+        if not self._gocert_initialized():
+            event.add_status(ops.BlockedStatus("Please initialize GoCert"))
+            return
+        event.add_status(ops.ActiveStatus())
+
+    def _on_gocert_notify(self, event: ops.PebbleCustomNoticeEvent):
```
Comment (gruyaume) [inline src/charm.py:104]: This likely deserves a docstring

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: charmcraft.yaml [hunk 6]
```diff
@@ -24,6 +24,21 @@ bases:
 containers:
   gocert:
     resource: gocert-image
+    mounts:
+    - storage: config
+      location: /etc/config
+    - storage: database
+      location: /etc/database
```
[the remark this answers is not in the document]
Comment (kayra1) [reply charmcraft.yaml:31]: where should I put it? I could just put everything under gocert: `/gocert/config` and `gocert/database`

File: src/charm.py (hunk 3 above)
Comment (kayra1) [reply src/charm.py:224]: if the file in the path given in the verify argument does not exist, this call will raise an OSError. it gets resolved during configuration when we create and save the ca.pem file.

File: src/charm.py (hunk 4 above)
Comment (kayra1) [reply src/charm.py:148]: The square brackets help me identify the important logs faster because they are easier to see when skimming the sea of logs. It was something we discussed in the pulse retro mentioned by @DanielArndt . I left it because i wanted to get everyone's opinion on it. If we like it i will add it to a custom logger object like we do in Vault, but if we don't I can remove it.

File: src/charm.py (hunk 5 above)
Comment (kayra1) [reply src/charm.py:104]: oops, the contents of this function should be in the configure_certificate_relations function, which has the docstring. this function is deleted.

File: src/charm.py (hunk 2 above)
Comment (kayra1) [reply src/charm.py:197]: I may just move the entire function to a different branch, because I want to have its tests in the same PR, but if I add those here this PR the size will increase a lot. is that ok?

Comment (kayra1) [reply src/charm.py]: I removed this logic from here and pushed to a [different branch](https://github.com/canonical/gocert-k8s-operator/pull/6).

File: charmcraft.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gruyaume) [reply charmcraft.yaml:31]: Typically under `/var/lib/<something>`, here that could be `/var/lib/gocert`

File: src/charm.py (hunk 4 above)
Comment (gruyaume) [reply src/charm.py:123]: I'm ok with changing the way we log but we should write a spec on it as it's quite important and affects every charm. I wouldn't include this change as part of something else though (like here).

File: src/charm.py [hunk 7]
```diff
@@ -95,6 +143,107 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+            content = CertificateSecret(
+                certificate=ca_cert,
+                private_key=ca_pk,
+            )
+        except ops.SecretNotFoundError:
+            pk = generate_private_key()
+            ca = generate_ca(private_key=pk, subject=SELF_SIGNED_CA_COMMON_NAME)
+            content = CertificateSecret(certificate=ca.decode(), private_key=pk.decode())
+            self.app.add_secret(label=SELF_SIGNED_CA_SECRET_LABEL, content=content.to_dict())
+        return content.certificate.encode(), content.private_key.encode()
```
Comment (gruyaume) [inline src/charm.py:245]: I'm not a huge fan that this `get_xyz` can create stuff, it feels like we should either change the function name or separate it in to.

File: src/charm.py [hunk 8]
```diff
@@ -95,6 +143,107 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+        except ops.pebble.PathError:
+            return False
+        try:
+            certificate = x509.load_pem_x509_certificate(existing_cert.read().encode())
+            common_name = certificate.issuer.get_attributes_for_oid(x509.NameOID.COMMON_NAME)[0]
+        except (ValueError, TypeError):
+            return False
+        if common_name.value != SELF_SIGNED_CA_COMMON_NAME:
+            logger.warning(common_name.value)
+            logger.warning(SELF_SIGNED_CA_COMMON_NAME)
```
Comment (gruyaume) [inline src/charm.py:225]: This makes for a bizare log doesn't it?

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 9]
```diff
@@ -95,6 +143,107 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+        if not binding.network:
+            return None
+        if not binding.network.bind_address:
+            return None
+        return str(binding.network.bind_address)
+
+    ## Status Checks ##
+    def _storages_attached(self) -> bool:
+        """Return if the storages are attached."""
+        return self.model.storages.get("config") and self.model.storages.get("database")  # type: ignore
```
Comment (gruyaume) [inline src/charm.py:160]: Why the type ignore here?

File: src/charm.py [hunk 10]
```diff
@@ -95,6 +143,107 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+            req = requests.get(
+                f"https://{self._application_bind_address}:{self.port}/status",
+                verify=f"{CHARM_PATH}/{CONFIG_MOUNT}/0/ca.pem",
+            )
+        except (requests.RequestException, OSError):
+            return False
+        if req.status_code != 200:
+            return False
+        body = req.json()
+        return body.get("initialized", False)
```
Comment (gruyaume) [inline src/charm.py:187]: This feels like we'll soon want to move out gocert api call to their own class

File: tests/unit/test_charm.py [hunk 11]
```diff
@@ -1,18 +1,128 @@
...
+
 import ops
 import ops.testing
 import pytest
+import requests
 from charm import GocertCharm
-from scenario import Context, Event, State
+from scenario import Container, Context, Event, Network, State, Storage
+
+# https://res.cloudinary.com/canonical/image/fetch/f_auto,q_auto/https://discourse-charmhub-io.s3.eu-west-2.amazonaws.com/original/2X/4/4ac42dc8a238a003c7d56fe282246ca102dd594f.png
```
Comment (gruyaume) [inline tests/unit/test_charm.py:14]: Was this addition intentional?

File: tests/unit/test_charm.py [hunk 12]
```diff
@@ -1,18 +1,128 @@
...
+            pytest.param(
+                Mock(
+                    **{"json.return_value": {"initialized": True}, "status_code": 200},
+                ),
+                "initialized",
+                id="gocert_initialized",
+            ),
+        ],
+    )
+    def test_configure_handler(
```
Comment (gruyaume) [inline tests/unit/test_charm.py:72]: Let's not change the convention we have for unit test names please.

File: tests/unit/test_charm.py [hunk 13]
```diff
@@ -1,18 +1,128 @@
...
 class TestCharm:
-    @pytest.fixture(scope="function", autouse=True)
+    @pytest.fixture(scope="function")
     def context(self):
         yield Context(GocertCharm)
 
-    def test_start_charm(self, context):
-        out = context.run(Event("start"), State())
-        assert out.unit_status == ops.ActiveStatus()
+    @pytest.mark.parametrize(
```
Comment (gruyaume) [inline tests/unit/test_charm.py:22]: This unit test is huge , are you trying to merge all tests into 1. It feels like we went too far here  :laughing:

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 10 above)
Comment (kayra1) [reply src/charm.py:187]: TLSENG-342

File: src/charm.py (hunk 8 above)
Comment (kayra1) [reply src/charm.py:225]: oops

File: src/charm.py (hunk 9 above)
Comment (kayra1) [reply src/charm.py:160]: For some reason, static checks don't believe that self.model.storages.get() returns a boolean value, even though if there's no storage it will return None.

I tried some different things, and casting them to bool seems to work so its removed.

File: tests/unit/test_charm.py (hunk 11 above)
Comment (kayra1) [reply tests/unit/test_charm.py:14]: I would have liked to keep it there, but moved this file to a different pr

File: tests/unit/test_charm.py (hunk 13 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: There is only 1 public function :D.

I parameterized all of the tests and moved this to a separate PR.

File: tests/unit/test_charm.py (hunk 12 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: reverted

File: charmcraft.yaml [hunk 14]
```diff
@@ -24,6 +24,21 @@ bases:
 containers:
   gocert:
     resource: gocert-image
+    mounts:
+    - storage: config
+      location: /var/lib/gocert/config
```
Comment (ghislainbourgeois) [inline charmcraft.yaml:29]: I think @gruyaume's comment was only for the DB, not the configuration. I think the config makes more sense in `/etc`

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: tests/unit/test_charm.py [hunk 15]
```diff
@@ -1,18 +1,674 @@
...
+            storage=[Storage(name="config")],
+            containers=[Container(name="gocert", can_connect=False)],
+            networks={"juju-info": Network([], [], [])},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+
+        assert out.unit_status == ops.WaitingStatus("container not yet connectable")
```
Comment (gruyaume) [inline tests/unit/test_charm.py:32]:
```suggestion
    def test_given_only_config_storage_container_cant_connect_network_not_available_gocert_not_running_when_collect_unit_status_then_status_is_waiting(
        self, context
    ):
        state = State(
            storage=[Storage(name="config")],
            containers=[Container(name="gocert", can_connect=False)],
            networks={"juju-info": Network([], [], [])},
            leader=True,
        )

        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
            out = context.run(Event("collect_unit_status"), state)

        assert out.unit_status == ops.WaitingStatus("container not yet connectable")
```

This comment applies to every collect status related test

File: tests/unit/test_charm.py [hunk 16]
```diff
@@ -1,18 +1,674 @@
...
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert len(out.secrets) == 0
+        root = out.containers[0].get_filesystem(context)
+        assert (root / "var/lib/gocert/config/config.yaml").open("r")
+        assert not (root / "var/lib/gocert/config/certificate.pem").exists()
+        assert not ((root / "var/lib/gocert/config/private_key.pem").exists())
+        assert out.unit_status == ops.WaitingStatus("certificates not yet created")
```
Comment (gruyaume) [inline tests/unit/test_charm.py:108]: I would avoid having a test that asserts both the content of files and unit status as those are handled by two different handlers ( though I know the collect unit status runs as an exit hook to other events).

File: tests/unit/test_charm.py [hunk 17]
```diff
@@ -1,18 +1,674 @@
...
+            .read()
+            .startswith("-----BEGIN CERTIFICATE-----")
+        )
+        assert (
+            (root / "var/lib/gocert/config/private_key.pem")
+            .open("r")
+            .read()
+            .startswith("-----BEGIN RSA PRIVATE KEY-----")
+        )
+        assert out.unit_status == ops.WaitingStatus("GoCert server not yet available")
```
Comment (gruyaume) [inline tests/unit/test_charm.py:208]: Same comment related to the status

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 18]
```diff
@@ -5,80 +5,128 @@
...
-                            certificate_signing_request=csr,
-                            certificate=cert,
-                            ca="",
-                            chain=[cert],
-                            relation_id=relation.id,
-                        )
+    ## Configure Dependencies ##
+    def _configure_gocert_config_file(self):
+        """Push the config file."""
+        logger.info("Configuring the config file.")
```
Comment (gruyaume) [inline src/charm.py:106]:
```suggestion
```

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: tests/unit/test_charm.py (hunk 15 above)
Comment (kayra1) [reply tests/unit/test_charm.py:32]: but these tests are also validating that the configure function does not crash. Should we not check that too?

Comment (gruyaume) [reply tests/unit/test_charm.py:32]: I think we want tests specifically for the collect unit status handler. Now we will also want tests for the configure/config-changed handlers. Those two are independent.

File: src/charm.py (hunk 7 above)
Comment (kayra1) [reply src/charm.py]: The docstring mentions that it will be created but i renamed the function too.

File: charmcraft.yaml (hunk 14 above)
Comment (kayra1) [reply charmcraft.yaml:29]: changed it back

File: tests/unit/test_charm.py [hunk 19]
```diff
@@ -1,18 +1,1264 @@
...
+    ):
+        state = State(
+            storage=[Storage(name="config")],
+            containers=[Container(name="gocert", can_connect=False)],
+            networks={"juju-info": Network([], [], [])},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
```
Comment (gruyaume) [inline tests/unit/test_charm.py:651]:
```suggestion
            out = context.run(Event("collect_unit_status"), state)
```

This comment applies to every test for collect unit status

File: tests/unit/test_charm.py [hunk 20]
```diff
@@ -1,18 +1,1264 @@
...
+        state = State(
+            storage=[Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network.default()},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("storages not yet available")
```
Comment (gruyaume) [inline tests/unit/test_charm.py:794]: I think we should parametrize tests that have the same assertions. For example I don't think the test for the db storage is available vs the config one can be combined into 1 (with parameters) as the expected behavior from our code is the same for both.

This may apply elsewhere, we probably don't want 3 different tests for `container not yet connectable`

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: tests/unit/test_charm.py (hunk 15 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: How are they independent if the collect unit status handler can't run without a preceding event handler? It doesn't make sense to me to have separate tests to run a function that will never run without another public function that is the actual entry point. It is practically no different than unit testing an internal function which we dont do most of the time.

Regardless, I will touch on this in the spec as it makes sense to talk about it there. I've moved all assert status checks into their own copy of the same test except we only run collect status there.

File: tests/unit/test_charm.py (hunk 16 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: changed

File: tests/unit/test_charm.py (hunk 17 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: changed

File: tests/unit/test_charm.py (hunk 19 above)
Comment (kayra1) [reply tests/unit/test_charm.py]: already changed

File: src/charm.py (hunk 18 above)
Comment (kayra1) [reply src/charm.py]: gone

File: src/charm.py [hunk 21]
```diff
@@ -5,80 +5,128 @@
...
+            logger.info("Config file already created.")
+        except ops.pebble.PathError:
+            config_file = open("src/config/config.yaml").read()
+            self.container.make_dir(path=f"{WORKLOAD_PATH}/config", make_parents=True)
+            self.container.push(path=f"{WORKLOAD_PATH}/config/config.yaml", source=config_file)
+            logger.info("Config file created.")
+
+    def _configure_access_certificates(self):
+        """Update the config files for gocert and replan if required."""
+        logger.info("Configuring certificates.")
```
[the remark this answers is not in the document]
Comment (kayra1) [reply src/charm.py]: gone

File: src/charm.py [hunk 22]
```diff
@@ -89,12 +137,113 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
...
+            return False
+        if req.status_code != 200:
+            return False
+        body = req.json()
+        return body.get("initialized", False)
+
+    ## Helpers ##
+    def _generate_self_signed_certificates(self) -> None:
+        """Generate self signed certificates and saves them to secrets and the charm."""
+        logger.info("Creating self signed certificates.")
```
[the remark this answers is not in the document]
Comment (kayra1) [reply src/charm.py]: gone

File: tests/unit/test_charm.py (hunk 20 above)
Comment (kayra1) [reply tests/unit/test_charm.py:794]: >I think we should parametrize tests that have the same assertions

I've tried doing this and in a short summary:
The issue with it is we have parameter combinations that will produce different results. This in turn results in us having to generalize the parameter to the entire state of the charm (a single parameter that includes the state of storage, containers, relations, workload etc.).

## There are 2 viable options to do parametrization:
1) The parameters are over the entire state of the charm like suggested, and include the expected assertion (it could be a status, a file name, or just straight up code)

The pros are:
* The tests look "cleaner" and easier to read since there is no branching
* The written tests to generated tests ratio is higher. This means there is less tests generated, which requires less reasoning about the parameters

The cons are:
* Requires the user to enumerate the entire state of the charm one by one. 

This is extremely similar to writing a new test for each parameter. We lose the very powerful parameter matrix ability of pytest that generates all possible combinations of the world state

* Since the world state is hand written, there will be gaps in the possibilities.
An outcome of the above con is the fact that we may skip over some world states when typing the parameters themselves that may turn out to be a problem (like storage not attached and gocert available. This was the cause of [this issue](https://github.com/canonical/vault-k8s-operator/issues/242) in Vault).

* It is very difficult to enumerate the state of the charm
Due to the nature of parameters in pytests, they all need to share the exact structure. We may want to enumerate a state that has storages, container and network binding, but if we do we must now include this in all enumerated parameters for the test. We will no longer have the option to just add a relation parameter, we must go back and add that field to every single param we described. This severely limits the benefits of having parameters, since copying and pasting the test is now more flexible than the parameter option.

2) The paremeters are over the possible states of each dependency like in pr #10 , and include their status in enumerated form
The pros are:
* We no longer have to think about the entire state of the world when developing the charm. 
When typing parameters, all we have to write is the dependency we want and its possible states. 
* The tests are more extensible
For example, all we need to do to test all of our handlers when we have a new relation added is to add a new parameter with 3 options: "Relation doesnt exist, only 1 exists, and multiple exist", and pytest will automatically generate all the new possible input states. Even if we don't add any assertions, the tests failing due to a raised error is immediately useful for charms since we do not want to raise anything in a charm.
* The files are much smaller
Even if the readability takes a hit, reading ~100 lines of parametrized tests end up being faster than reading ~1000 lines of code with many duplicates. I personally end up skimming and miss crucial minor differences when writing tests.
* the tests are fully exhaustive.
With this method, tests are exhaustive by default, and doesn't require the user to think about the states of the world at all. Just write assertions for the happy path, and trust that no error will be raised when the charm is deployed with any combination of relations, storages, connections and stages.

The cons are:
* It is a different way of reading and writing tests, which requires a change in thinking and reading
We are used to reading the title format that we have, and easily validate that the test does what the title says. If the title doesn't contain some dependency like a `peer-relation`, it is common that we do not check if that test initializes this relation because we cant know just by looking at it that it would require it. In the proposed way, we would not have to pay attention to these dependencies, and trust that as long as the parameter is exhaustive, we will test it in every possible way.

When reading and trying to understand what a test does, we now have to focus on the possible output states are instead, which aligns well with how we think about writing charms. 

* The test itself will have a bunch of branches which make it harder to read.
When evaluating this con, I frequently found myself thinking about the ratio of expected output states and the possible input states. The smaller the ratio, the better the proposed method becomes. In the case of gocert at this point, this ratio is about 7/72: 7 output states to 72 different input states, and the input states grow exponentially when the output states grow linearly. This function also bodes well to how well the approach scales as the charm grows in functionality. When a single relation and the connection to the model is added, this ratio shrinks to ~15/432. Since we also have no problem reading long branching functions in places like the `_configure` functions, I feel that we as a team are in a better position to read these types of tests.

There is actually a way to combine the two ways with pytest's metaprogramming features, but frankly that is too much for now.

all of this and examples will be a part of the spec

Comment (gruyaume) [reply tests/unit/test_charm.py:794]: Thank you for putting the effort into this, let's discuss this in the spec

**Review by gruyaume — APPROVED** [review]

**Review by ghislainbourgeois — APPROVED** [review]
